plutus-benchmark/src/Plutus/Benchmark/Queens.hs

```diff
@@ -0,0 +1,553 @@
+
+{-# INLINABLE head #-}
+head :: [a] -> a
+head (x:_) =  x
+head []    =  TxPrelude.error ()
+
+{-# INLINABLE tail #-}
+tail :: [a] -> [a]
+tail (_:as) =  as
+tail []     =  TxPrelude.error ()
+
+infix 4 `notElem`
+
+{-# INLINABLE notElem #-}
+notElem :: (Eq a) => a -> [a] -> Bool
+notElem a as = not (a `elem` as)
+
+-- Replacement for [a..b]
+{-# INLINABLE interval #-}
+interval :: Integer -> Integer -> [Integer]
+interval a b =
+    if a > b then []
+    else a : (interval (a+1) b)
+
+{-# INLINABLE abs #-}
+abs :: Integer -> Integer
+abs n = if n<0 then 0-n else n
+
+-- Things needed for `union`
+
+{-# INLINABLE elem_by #-}
+elem_by :: (a -> a -> Bool) -> a -> [a] -> Bool
+elem_by _  _ []     =  False
+elem_by eq y (x:xs) =  x `eq` y || elem_by eq y xs
+
+{-# INLINABLE nubBy #-}
+nubBy :: (a -> a -> Bool) -> [a] -> [a]
+nubBy eq l              = nubBy' l []
+  where
+    nubBy' [] _         = []
+    nubBy' (y:ys) xs
+       | elem_by eq y xs = nubBy' ys xs
+       | otherwise       = y : nubBy' ys (y:xs)
+
+{-# INLINABLE deleteBy #-}
+deleteBy :: (a -> a -> Bool) -> a -> [a] -> [a]
+deleteBy _  _ []     = []
+deleteBy eq x (y:ys) = if x `eq` y then ys else y : deleteBy eq x ys
+
+{-# INLINABLE unionBy #-}
+unionBy :: (a -> a -> Bool) -> [a] -> [a] -> [a]
+unionBy eq xs ys =  xs ++ foldl (flip (deleteBy eq)) (nubBy eq ys) xs
+
+{-# INLINABLE union #-}
+union :: (Eq a) => [a] -> [a] -> [a]
+union                   = unionBy (==)
+
+-- Things needed for `sortBy`
+instance Eq Ordering where
+    LT == LT = True
+    EQ == EQ = True
+    GT == GT = True
+    _ == _ = False
+
+{-# INLINABLE sortBy #-}
+--- Stolen from Data.List
+sortBy :: (a -> a -> Ordering) -> [a] -> [a]
+sortBy cmp = mergeAll . sequences
+  where
+    sequences (a:b:xs)
+      | (a `cmp` b) == GT = descending b [a]  xs
+      | otherwise       = ascending  b (a:) xs
+    sequences xs = [xs]
+
+    descending a as (b:bs)
+      | (a `cmp` b) == GT = descending b (a:as) bs
+    descending a as bs  = (a:as): sequences bs
+
+    ascending a as (b:bs)
+      | (a `cmp` b) /= GT = ascending b (\ys -> as (a:ys)) bs
+    ascending a as bs   = case as [a]  -- Original used let !x = ...
+                          of x -> x : sequences bs
+
+    mergeAll [x] = x
+    mergeAll xs  = mergeAll (mergePairs xs)
+
+    mergePairs (a:b:xs) = case merge a b -- Original used let !x = ...
+                          of x ->  x : mergePairs xs
+    mergePairs xs       = xs
+
+    merge as@(a:as') bs@(b:bs')
+      | a `cmp` b == GT = b:merge as  bs'
+      | otherwise       = a:merge as' bs
+    merge [] bs         = bs
+    merge as []         = as
+
+
+-----------------------------
+-- Figure 1. CSPs in Haskell.
+-----------------------------
+
+type Var = Integer
+type Value = Integer
+
+data Assign = Var := Value deriving (Show)
+instance Eq Assign
+    where (a := b) == (a' := b') = a==a' && b == b'
+
+instance Ord Assign
+    where (a := b) < (a' := b') = (a<a') || (a==a' && b < b')
```

Comment:
Can you just derive those? Or do we not have deriving in Plutus Tx? But `deriving (Show)` works?

Reply:
I'm not sure.  The original program has 

```
data Assign = Var := Value deriving (Eq, Ord, Show)
```

but for some reason I had to write my own instances.